### PR TITLE
Fix #61: Add support callable middlewares like `[$instance, 'method']` and `$invokeable`

### LINKED ISSUE #61: Add support callable middlewares, not `Closure` only

### PR DESCRIPTION
**Pull request by rustamwin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️/❌
| Fixed issues  | #61

### CODE REVIEW

File: src/MiddlewareFactory.php [hunk 1]
```diff
@@ -44,8 +44,9 @@ public function __construct(
      * - A callable with
      *   `function(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface`
      *   signature.
-     * - A controller handler action in format `[TestController::class, 'index']`. `TestController` instance will
-     *   be created and `index()` method will be executed.
+     * - A callable object instance.
```
Comment (vjik) [inline src/MiddlewareFactory.php:47]:
```suggestion
     * - Any callable.
```

File: src/MiddlewareFactory.php [hunk 2]
```diff
@@ -87,21 +88,21 @@ private function isMiddlewareClassDefinition(array|callable|string $definition):
     }
 
     /**
-     * @psalm-assert-if-true array{0:class-string, 1:non-empty-string}|Closure $definition
+     * @psalm-assert-if-true array{0:class-string|object, 1:non-empty-string}|callable-object|Closure $definition
      */
     private function isCallableDefinition(array|callable|string $definition): bool
     {
-        if ($definition instanceof Closure) {
+        if ($definition instanceof Closure || is_object($definition)) {
```
Comment (vjik) [inline src/MiddlewareFactory.php:95]:
```suggestion
        if (is_callable($definition)) {
```

Object may not be callable.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/MiddlewareFactory.php (hunk 2 above)
Comment (rustamwin) [reply src/MiddlewareFactory.php:95]: Hmm... the method signature only allows the callable object?

Comment (vjik) [reply src/MiddlewareFactory.php:95]: Not only. Any callable: `[$object, 'method']`, `'anyFunctionName'`, object with `__invoke`, etc.

**Review by samdark — APPROVED** [review]

File: src/MiddlewareFactory.php [hunk 3]
```diff
@@ -87,21 +88,21 @@ private function isMiddlewareClassDefinition(array|callable|string $definition):
...
     {
-        if ($definition instanceof Closure) {
+        if (is_callable($definition)) {
             return true;
         }
 
         return is_array($definition)
             && array_keys($definition) === [0, 1]
-            && is_string($definition[0])
+            && (is_string($definition[0]) || is_object($definition[0]))
```
Comment (vjik) [inline src/MiddlewareFactory.php:101]: It's not need, because `is_callable` above cover this case.

File: src/MiddlewareFactory.php [hunk 4]
```diff
@@ -87,21 +88,21 @@ private function isMiddlewareClassDefinition(array|callable|string $definition):
...
             && is_string($definition[1])
             && in_array(
                 $definition[1],
-                class_exists($definition[0]) ? get_class_methods($definition[0]) : [],
+                is_object($definition[0]) || class_exists($definition[0]) ? get_class_methods($definition[0]) : [],
```
Comment (vjik) [inline src/MiddlewareFactory.php:105]: It's not need, because `is_callable` above cover this case.

File: src/MiddlewareFactory.php [hunk 5]
```diff
@@ -125,9 +126,9 @@ private function isArrayDefinition(array|callable|string $definition): bool
     }
 
     /**
-     * @param array{0:class-string, 1:non-empty-string}|Closure $callable
+     * @param array{0:class-string|object, 1:non-empty-string}|callable-object|callable-string|Closure $callable
```
Comment (vjik) [inline src/MiddlewareFactory.php:129]:
```suggestion
     * @param array{0:class-string, 1:non-empty-string}|callable|Closure $callable
```

File: src/MiddlewareFactory.php [hunk 6]
```diff
@@ -184,17 +185,17 @@ private function getCallableParameters(): array
     }
 
     /**
-     * @param class-string $class
+     * @param class-string|object $class
```
Comment (vjik) [inline src/MiddlewareFactory.php:188]: It's not need, because `is_callable` above cover this case.

File: src/MiddlewareFactory.php [hunk 7]
```diff
@@ -184,17 +185,17 @@ private function getCallableParameters(): array
...
      * @param non-empty-string $method
      */
-    private function createActionWrapper(string $class, string $method): MiddlewareInterface
+    private function createActionWrapper(string|object $class, string $method): MiddlewareInterface
```
Comment (vjik) [inline src/MiddlewareFactory.php:191]: It's not need, because `is_callable` above cover this case.

File: src/MiddlewareFactory.php [hunk 8]
```diff
@@ -184,17 +185,17 @@ private function getCallableParameters(): array
...
     {
         return new class ($this->container, $this->parametersResolver, $class, $method) implements MiddlewareInterface {
             public function __construct(
                 private ContainerInterface $container,
                 private ?ParametersResolverInterface $parametersResolver,
-                /** @var class-string */
-                private string $class,
+                /** @var class-string|object */
+                private string|object $class,
```
Comment (vjik) [inline src/MiddlewareFactory.php:198]: It's not need, because `is_callable` above cover this case.

File: src/MiddlewareFactory.php [hunk 9]
```diff
@@ -204,8 +205,12 @@ public function process(
...
                 RequestHandlerInterface $handler
             ): ResponseInterface {
-                /** @var mixed $controller */
-                $controller = $this->container->get($this->class);
+                if (is_string($this->class)) {
+                    /** @var mixed $controller */
+                    $controller = $this->container->get($this->class);
+                } else {
+                    $controller = $this->class;
+                }
```
Comment (vjik) [inline src/MiddlewareFactory.php:213]: It's not need, because `is_callable` above cover this case.

**Review by samdark — APPROVED** [review]

File: src/MiddlewareFactory.php [hunk 10]
```diff
@@ -44,8 +44,9 @@ public function __construct(
      * - A callable with
      *   `function(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface`
      *   signature.
-     * - A controller handler action in format `[TestController::class, 'index']`. `TestController` instance will
-     *   be created and `index()` method will be executed.
+     * - Any callable.
+     * - A controller handler action in format `[TestController::class, 'index']` or `[new TestController(), 'index']`.
+     *   `TestController` instance will be created if it is not an instance and `index()` method will be executed.
```
Comment (vjik) [inline src/MiddlewareFactory.php:49]:
```suggestion
     * - A controller handler action in format `[TestController::class, 'index']`. `TestController` instance will
     *   be created and `index()` method will be executed.
```

**Review by vjik — APPROVED** [review]

**Pull request merged**